src/System.Management.Automation/engine/NativeCommandParameterBinder.cs

```diff
@@ -324,12 +340,76 @@ private void PossiblyGlobArg(string arg, bool usedQuotes)
             }
 #endif // UNIX
 
+            // Check if the start of the path is a filesystem drive
+            if (stringConstantType != StringConstantType.SingleQuoted)
+            {
+                arg = ResolvePath(arg, Context);
+            }
+
             if (!argExpanded)
             {
                 _arguments.Append(arg);
             }
         }
 
+        /// <summary>
+        /// Check if string is prefixed by psdrive, if so, expand it if filesystem path.
+        /// </summary>
+        /// <param name="path">The potential PSPath to resolve.</param>
+        /// <param name="context">The current ExecutionContext.</param>
+        /// <returns>Resolved PSPath if applicable otherwise the original path</returns>
+        internal static string ResolvePath(string path, ExecutionContext context)
+        {
+            if (ExperimentalFeature.IsEnabled("PSNativePSPathResolution"))
+            {
+                string driveName;
+#if !UNIX
+                if (string.Equals(path, "~") || path.StartsWith("~" + Path.DirectorySeparatorChar) || path.StartsWith("~" + Path.AltDirectorySeparatorChar))
+                {
+                    try
+                    {
+                        ProviderInfo fileSystemProvider = context.EngineSessionState.GetSingleProvider(FileSystemProvider.ProviderName);
+                        return (fileSystemProvider.Home + path.Substring(1)).Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
```

Comment:
```suggestion
                        return new StringBuilder(fileSystemProvider.Home) // Is this guaranteed to end with a slash?
                            .Append(path.Substring(1))
                            .Replace(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar) // Note that on Linux and macOS this replaces '/' with '/' and I'm not sure if the method will detect that to be a no-op
                            .ToString();
```

Reply:
This way we can reuse the same buffer for replacement without reallocating it before we turn it into a string

Reply:
Looks like [there's no check for replacing a char with itself](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs#L2025-L2074). [There is one for characters in strings](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L1055-L1056).

Reply:
We could replace StringBuilder class with string.Create() method.

Reply:
This code only applies to !UNIX anyways.  Will update to use string builder.